Add conversion script for drs4 pedestal file to EVB format

maxnoe opened the pull request.

FrancaCassol commented:
Hi @maxnoe, 
I have two questions concerning this script: 
1) The best it would be that the EvB read directly the h5 file, could you please recall why this conversion step is necessary? 
2) why did you prefer a script to a tool ?

maxnoe commented:
> The best it would be that the EvB read directly the h5 file, could you please recall why this conversion step is necessary?

EVB does not depend on `libhdf5` right now and it would be a major new dependency for them. I don't think they are keen on doing that.

> why did you prefer a script to a tool ?

There is nothing to configure here. We don't use anything using the ctapipe configuration system (no components) so it's just less code.

FrancaCassol commented:
> EVB does not depend on `libhdf5` right now and it would be a major new dependency for them. I don't think they are keen on doing that. 

mmh, before to double  the number of pedestal files, I suggest to ask them directly.

In any case, if we decide to go in this direction the two formats should contain exactly the same data, so the round at integer should be done before. 
Also, the provenance info is transmitted only for the h5 file, if the binary file is that used by the EvB, it  should be collected also for it.

> It's just that much simpler if we provide the file in the format they need. (And no changes to EVB required at all)

EVB is not finalised in this part, so some changes, if acceptable, can be proposed. The price to pay here is to double the number of files and to complicate the book-keeping

maxnoe commented:
> mmh, before to double the number of pedestal files

I don't think we should create the binary file for each night. Just after hardware interventions when EVB needs a new file.

> , I suggest to ask them directly.

I did. They requested the format stays the same (same discussion with fits actually, cfitsio would even be simpler to have as a depencency than hdf5).

> In any case, if we decide to go in this direction the two formats should contain exactly the same data, so the round at integer should be done before.

I disagree. We have a much more flexible data format, so we can include more data (e.g. the counts and std devs).

> Also, the provenance info is transmitted only for the h5 file, if the binary file is that used by the EvB, it should be collected also for it.

That's just not possible with the EVB format. It's just a bag of bytes. There is no way to add any provenance data to the file itself.

maxnoe commented:
@DirkHoffman just confirmed to me again, that this is as it should be from the EVB maintainer side.

FrancaCassol commented:
> I don't think we should create the binary file for each night. Just after hardware interventions when EVB needs a new file.

Yes, they must be create each night (the baseline slightly changes at each camera power cycle), so we will have two files per night, containing different data and of which the one used by the EvB not tracked by provenance 

> That's just not possible with the EVB format. It's just a bag of bytes. There is no way to add any provenance data to the file itself.

then (a least) we must add inside the h5 format the data given to the EvB. 

> Adding a dependency to EVB (either libhdf5 or cfitsio) is also a burden and probably much larger than maintaining this simple conversion script and running it after a hardware intervention.

The script is not the problem, the problem is the doubling of data and book-keeping

FrancaCassol commented:
> @DirkHoffman just confirmed to me again, that this is as it should be from the EVB maintainer side.

For a clear decision book-keeping , I think it is necessary an official statement from the EvB people (@houlesjulien and @DirkHoffmann ) that the EvB can not read h5 files without strong constraining to the CameraServer. Then, we must obviously go for the two formats solution, but guaranteeing the data traceability.

maxnoe commented:
@FrancaCassol no, it is not. You are requesting a change of already made decisions. I completely agree with them, that this is not worth it.

> Yes, they must be create each night (the baseline slightly changes at each camera power cycle), so we will have two files per night, containing different data and of which the one used by the EvB not tracked by provenance

Since when are the files used by EVB for the pre-calibration changed each night? Or when is this change planned to happen if it was not made yet?

> then (a least) we must add inside the h5 format the data given to the EvB.

No. That is completely unnecessary, since there is a clear mapping between the values. There is no point in storing data twice in two files if you have a clearly defined a way of getting the one from the other.

> The script is not the problem, the problem is the doubling of data and book-keeping

There is no double book keeping. It The binary file for EVB is a clear product of the hdf5 file. 


I quote wrote Dirk wrote to me in Slack (translated from original German):

> The interfaces of EVB are all fixed and were confirmed in two CDRs. To touch that again, there has to be a formal Change Request. [...] Until CTAO defines the official formats, no change in data format is on the agenda.

moralejo commented:
Hi Franca, by official you mean that @houlesjulien or @DirkHoffmann make statement in this thread? 
I can understand they want to avoid adding a dependency - considering also we have a working format.

Doubling of data is not a major issue considering the absolute size, right?  As for traceability, can you make a proposal? Is it enough to add in the h5 the exact numbers that are written in the file for evb?  Or do you also want to add something in the evb file? (e.g. change the format to contain the name & checksum of the original h5)

maxnoe commented:
> (e.g. change the format to contain the name & checksum of the original h5)

The only place to put that would be the filename or a text file next to it, everything else would be a change to the contents of the file itself.

moralejo commented:
> 
> Since when are the files used by EVB for the pre-calibration changed each night? Or when is this change planned to happen if it was not made yet?
> 

This is anyway unrelated to the format discussion, whatever format would allow to change the calibration on a nightly basis if wanted. And I think in the final standard operation (don't know how far is that) this  will be the case, since we should provide in real-time well-calibrated waveforms not to be re-calibrated again, and that is the best way to guarantee it.

FrancaCassol commented:
> Hi Franca, by official you mean that @houlesjulien or @DirkHoffmann make statement in this thread? I can understand they want to avoid adding a dependency - considering also we have a working format.

Yes, I think this is enough to go for it. 

> Doubling of data is not a major issue considering the absolute size, right? As for traceability, can you make a proposal? Is it enough to add in the h5 the exact numbers that are written in the file for evb? Or do you also want to add something in the evb file? (e.g. change the format to contain the name & checksum of the original h5)

I think provenance is the official method to guarantee traceability, so I would say that is enough to add in the h5 file the precise data to be used by the EvB. Then, the translation to the binary file could also be done on-the-fly in a temporary directory

maxnoe commented:
> I think provenance is the official method to guarantee traceability, so I would say that is enough to add in the h5 file the precise data to be used by the EvB. Then, the translation to the binary file could also be done on-the-fly in a temporary directory

This is already the case. It is by no means required to have the data as rounded ints in the hdf5 for that. I don't really care how this script is used in the end. It should be added to version control here, so it can be tested together with the program creating the pedestal file. Everything else is a question for osa or whatever is producing the nightly files in the end.

FrancaCassol commented:
> 
> Since when are the files used by EVB for the pre-calibration changed each night? Or when is this change planned to happen if it was not made yet?
 
All we are doing now is for the future. I guess the change is foreseen for spring 2022, but it depends on several developments (among others, this one)

> > then (a least) we must add inside the h5 format the data given to the EvB.

> No. That is completely unnecessary, since there is a clear mapping between the values. There is no point in storing data twice in two files if you have a clearly defined a way of getting the one from the other.

What is the problem with having the round values in the h5?

> Let me quote what wrote Dirk wrote to me in Slack (translated from original German):
> 
> > The interfaces of EVB are all fixed and were confirmed in two CDRs. To touch that again, there has to be a formal Change Request. [...] Until CTAO defines the official formats, no change in data format is on the agenda.

Perfect, then it is enough to have here a confirmation of the necessity to keep the binary file and the reference to those decisions.

houlesjulien commented:
Dear all, 

the script was originally written to replace Caco in providing the pedestal information. But if I understood correctly, Caco isn't made to manage this. 
The content of the file, not the file itself, was originally a parameter, amongst many others, for the transition to READY. We had to add a reader of a file because it wasn't implemented in Caco. 
Do we actually now how those information will be transmitted to EVB in the future ? 
If not, I propose to keep the system working as it works now. When a decision is taken, we will study it. 
Unfortunately, I won't have time to work on it. So if a change is requested, it would be more efficient if someone could provide a function written in C translating any format chosen into the one used in EVB. 

Best, 

Julien. 




De: "Franca Cassol" ***@***.***> 
À: "cta-observatory/cta-lstchain" ***@***.***> 
Cc: "houlesjulien" ***@***.***>, "Mention" ***@***.***> 
Envoyé: Mercredi 5 Janvier 2022 14:37:36 
Objet: Re: [cta-observatory/cta-lstchain] Add conversion script for drs4 pedestal file to EVB format (PR #829) 







Since when are the files used by EVB for the pre-calibration changed each night? Or when is this change planned to happen if it was not made yet? 



All we are doing now is for the future. I guess the change is foreseen for spring 2022, but it depends on several developments (among others, this one) 
BQ_BEGIN

BQ_BEGIN


then (a least) we must add inside the h5 format the data given to the EvB. 
BQ_END

BQ_END

BQ_BEGIN


No. That is completely unnecessary, since there is a clear mapping between the values. There is no point in storing data twice in two files if you have a clearly defined a way of getting the one from the other. 
BQ_END


What is the problem with having the round values in the h5? 
BQ_BEGIN


Let me quote what wrote Dirk wrote to me in Slack (translated from original German): 
BQ_BEGIN


The interfaces of EVB are all fixed and were confirmed in two CDRs. To touch that again, there has to be a formal Change Request. [...] Until CTAO defines the official formats, no change in data format is on the agenda. 
BQ_END

BQ_END


Perfect, then it is enough to have here a confirmation of the necessity to keep the binary file and the reference to those decisions. 

— 
Reply to this email directly, [ https://github.com/cta-observatory/cta-lstchain/pull/829#issuecomment-1005692308 | view it on GitHub ] , or [ https://github.com/notifications/unsubscribe-auth/AXFI2SUZ3BU7KNVCRW6M57LUURCSBANCNFSM5KI4KIGA | unsubscribe ] . 
Triage notifications on the go with GitHub Mobile for [ https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675 | iOS ] or [ https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub | Android ] . 
You are receiving this because you were mentioned. Message ID: ***@***.***>

maxnoe commented:
> So if a change is requested, it would be more efficient if someone could provide a function written in C translating any format chosen into the one used in EVB.

I thought EVB is C++?

houlesjulien commented:
No, it is implemented in C. 


De: "Maximilian Nöthe" ***@***.***> 
À: "cta-observatory/cta-lstchain" ***@***.***> 
Cc: "houlesjulien" ***@***.***>, "Mention" ***@***.***> 
Envoyé: Mercredi 5 Janvier 2022 16:40:22 
Objet: Re: [cta-observatory/cta-lstchain] Add conversion script for drs4 pedestal file to EVB format (PR #829) 







So if a change is requested, it would be more efficient if someone could provide a function written in C translating any format chosen into the one used in EVB. 



I thought EVB is C++? 

— 
Reply to this email directly, [ https://github.com/cta-observatory/cta-lstchain/pull/829#issuecomment-1005835571 | view it on GitHub ] , or [ https://github.com/notifications/unsubscribe-auth/AXFI2SVU7AENW7K56DZNPW3UURQ6NANCNFSM5KI4KIGA | unsubscribe ] . 
Triage notifications on the go with GitHub Mobile for [ https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675 | iOS ] or [ https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub | Android ] . 
You are receiving this because you were mentioned. Message ID: ***@***.***>

FrancaCassol commented:
So, as far as I understood, the situation is that the communication between Caco and the EvB is not yet defined (and I do not think it is to us to define it). Personally, what I care here is that the data delivered by the calibration system are traced in provenance and correspond strictly to the data asked as input by the EvB. At present, the h5 file are our (traced) deliverable. Then, if the EvB accepts floats values as input for the pedestals, we are fine with the present h5 files (the translation script will not belong the the calibration system, it can be considered as a tool for the EvB), but if the EvB expects integers, we need to have integers inside the h5 file (that, I repeat, is our traced deliverable). I had a look in the EvB ICD document, I could find a description of the pedestal table, but not of the type of data, @houlesjulien could you please confirm which type of data the EvB expects to receive for what concerns the baseline values?

houlesjulien commented:
EVB needs unsigned short values. 




De: "Franca Cassol" ***@***.***> 
À: "cta-observatory/cta-lstchain" ***@***.***> 
Cc: "houlesjulien" ***@***.***>, "Mention" ***@***.***> 
Envoyé: Jeudi 6 Janvier 2022 13:18:39 
Objet: Re: [cta-observatory/cta-lstchain] Add conversion script for drs4 pedestal file to EVB format (PR #829) 





So, as far as I understood, the situation is that the communication between Caco and the EvB is not yet defined (and I do not think it is to us to define it). Personally, what I care here is that the data delivered by the calibration system are traced in provenance and correspond strictly to the data asked as input by the EvB. At present, the h5 file are our (traced) deliverable. Then, if the EvB accepts floats values as input for the pedestals, we are fine with the present h5 files (the translation script will not belong the the calibration system, it can be considered as a tool for the EvB), but if the EvB expects integers, we need to have integers inside the h5 file (that, I repeat, is our traced deliverable). I had a look in the EvB ICD document, I could find a description of the pedestal table, but not of the type of data, [ https://github.com/houlesjulien | @houlesjulien ] could you please confirm which type of data the EvB expects to receive for what concerns the baseline values? 

— 
Reply to this email directly, [ https://github.com/cta-observatory/cta-lstchain/pull/829#issuecomment-1006538281 | view it on GitHub ] , or [ https://github.com/notifications/unsubscribe-auth/AXFI2STTHRAOE3DU2DYN3GDUUWCB7ANCNFSM5KI4KIGA | unsubscribe ] . 
Triage notifications on the go with GitHub Mobile for [ https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675 | iOS ] or [ https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub | Android ] . 
You are receiving this because you were mentioned. Message ID: ***@***.***>

moralejo commented:
So, is it ok for everyone to change the precision in the h5 file to 1 count? The difference is bound to be irrelevant for performance (given what 1 adc count is in terms of p.e., and comparing it to the poissonian noise), so I am fine with it.

FrancaCassol commented:
@maxnoe could you please update this PR in order to be coherent with PR #848 ? Then we can merge both

maxnoe commented:
@FrancaCassol let's merge #848 first, so this can be properly tested here.

FrancaCassol commented:
> @FrancaCassol let's merge #848 first, so this can be properly tested here.

@rlopezcoto?

maxnoe commented:
After merging #848, I replaced the rounding / convert to uint16 by a check that the pedestal is in the correct dtype. The script prints an error and exits with errorcode 1 if it is not the case.

I updated the test calibration files on the data server and the paths used here, so the test is passing.
